plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -280,28 +280,9 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
             }
          } ) );
 
-         // since the watermark has to be set before a block is created, we are looking into the future to
-         // determine the new schedule to identify producers that have become active
-         chain::controller& chain = chain_plug->chain();
-         const auto hbn = bsp->block_num;
-         auto new_pbhs = bsp->next(bsp->header.timestamp.next(), 0);
-
-         // for newly installed producers we can set their watermarks to the block they became active
-         if( bsp->active_schedule.version != new_pbhs.active_schedule.version ) {
-            flat_set<account_name> new_producers;
-            new_producers.reserve(new_pbhs.active_schedule.producers.size());
-            for( const auto& p: new_pbhs.active_schedule.producers) {
-               if (_producers.count(p.producer_name) > 0)
-                  new_producers.insert(p.producer_name);
-            }
-
-            for( const auto& p: bsp->active_schedule.producers) {
-               new_producers.erase(p.producer_name);
-            }
-
-            for (const auto& new_producer: new_producers) {
-               _producer_watermarks[new_producer] = hbn;
-            }
+         // simplify handling of watermark in on_block
+         if (bsp->block) {
+            _producer_watermarks[bsp->block->producer] = bsp->block_num;
```

Comment:
You don't need the extra indirection to the through the `block` shared pointer to get to the producer. You can access it via `bsp->header.producer`. This would also work if `bsp->block` isn't present, although I suppose that should never be the case in this context.

This only needs to be done for producers that are configured on the local node. So only if `bsp->header.producer` can be found inside `_producers`.